docs/src/developer_docs/model_interface.md

```diff
@@ -0,0 +1,227 @@
+# Model interface
+
+Oceananigans models are concrete subtypes of `AbstractModel` that can be advanced
+by [`Simulation`](@ref Simulation). This page documents the *model interface*:
+the handful of functions and data that a model must provide so that Simulation
+can initialize it, march it forward, and coordinate callbacks, diagnostics, and
+output writers.
+
+The interface lives primarily inside `src/Simulations/` and is intentionally
+minimal so that new models (or pedagogical toy models) can be written without
+depending on the full ocean model infrastructure.
+
+## Lifecycle overview
+
+When `run!(sim::Simulation)` is called the following high-level sequence occurs:
+
+1. `initialize!(sim.model)` prepares the model state,
+   etc.) and `update_state!(sim.model)` computes any auxiliary tendencies.
+
+2. Time-stepping begins. For every time step, Simulation computes an aligned `Δt`, gathers callbacks
+   that should run inside the model (`ModelCallsite`s), and calls
+   `time_step!(sim.model, Δt; callbacks=model_callbacks)`. Most models will also call `update_state!`
+   at the end of `time_step!`. This ensures that the diagnostic state is current with the prognostic state
+   so that output and callbacks can execute correctly on a fully consistent model state.
+
+3. After `time_step!(model, Δt)`, `Simulation` executes diagnostics, output
+   writers, and callbacks scheduled on the `TimeStepCallsite`.
+
+Because `Simulation` assumes this protocol, any custom `AbstractModel` should
+implement (or inherit sane fallbacks for) the items listed below.
+
+## Structure and extensions of an `AbstractModel`
+
+### Required properties
+
+- `model.grid`: Simulation uses the grid to determine `architecture(model)` 
+  and `eltype(model)` via the fallbacks `architecture(model) = model.grid.architecture`
+  and `eltype(model) = eltype(model.grid)`.
+
+- `model.clock :: Clock`: the source of truth for `time(model)` and
+  `iteration(model)`. Simulation uses it for stop criteria and logging, and
+  resets it via `reset_clock!(model)` when `reset!(sim)` is called.
+
+### Lifecycle hooks
+
+- `update_state!(model, callbacks=[]; compute_tendencies=true)`: invoked by
+  Simulation right after `initialize!` and inside most time steppers. This is
+  where models fill halos, update boundary conditions, recompute auxiliary
+  fields, and run callbacks with an `UpdateStateCallsite`. Implementations
+  typically finish by calling `compute_tendencies!(model, callbacks)` so that
+  any `TendencyCallsite` callbacks can modify tendencies before integration.
```

Comment:
is there documentation for this? can we link to it here?

Reply:
There is documentation here: https://clima.github.io/OceananigansDocumentation/stable/appendix/library/#Oceananigans.Simulations.Callback

Note that the way callbacks are executed via `TendencyCallsite` and `UpdateStateCallsite` is actually a property of the model design. Only `TimeStepCallsite` is owned by `Simulation`.